### PR TITLE
Durable contracts

### PR DESCRIPTION
**Pull request by Chris-Hibbert**
closes: #1126

## Description

Collect the details on creating durable contracts in one place. Update to cover use of `meta`  with `start` and stop recommending `prepare` for this.

Also provide some cross-linking between various descriptions of the approach.

### CODE REVIEW

**Comment by Chris-Hibbert** [comment]
Lint is failing and claiming there are warnings for two files. When I run `yarn prettier` locally, it doesn't find any issues.

File: main/guides/zoe/contract-details.md [hunk 1]
```diff
@@ -0,0 +1,121 @@
+# Durable Contract Details
+
+Zoe allows you to write smart contracts that manage interactions between cooperative, but possibly untrusting,
+parties. Some contracts perform a single simple task like trading one good for another, and then are
+done. Others stay around and manage assets and interactions over time. The first kind don't need persistence
+or the ability to upgrade, while the second kind need to make use of more sophisticated functionality.
+
+When a contract is intended to continue running and serving many customers over a long time, its objects and
+data need to be persistent, its owners or managers may need to be able to adjust parameters or perform other
+governance actions, and it needs to be upgradeable so that the code can adapt over time.
```
Comment (dckc) [inline main/guides/zoe/contract-details.md:10]: "needs"? as in _every_ long-running multi-customer contract _must_ be upgradable?

Immutable contracts are ideal in some ways, no?

File: main/guides/zoe/contract-details.md [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+`exoClassKit()`. `init`'s return value is the state that will be preserved by SwingSet for each
+object. `exo()` doesn't have an init parameter, so it creates and returns the singleton immediately rather
+than returning a maker.
+
+The Exo objects (or just "Exos") defined by these functions are both persistent and virtualizable. SwingSet
+knows their complete state, so it can page them out when space is tight, and page them back in when they are
+referenced again.
+
+Fields of the object returned by the `init` function become fields of the persistent state. (These cannot
+currently be changed on upgrade, though we're considering relaxing that restriction.) Within methods they can
```
Comment (dckc) [inline main/guides/zoe/contract-details.md:62]:
```suggestion
Fields of the object returned by the `init` function become fields of the persistent state. (_These cannot
currently be changed on upgrade, though we're considering relaxing that restriction._) Within methods they can
```

File: main/guides/zoe/contract-details.md [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+each be accessed as fields of `this.state`. Our convention is to extract the fields that will be used in a
+method on its first line, like `const { a, b } = this.state;` Once that has been done, those variable can be
+read or written as normal javascript variables, and the values will persist. (For more details, see [the note
+here](https://docs.agoric.com/guides/zoe/contract-upgrade.html#kinds)).
+
+### Methods: defining behavior
+
+The methods argument of `exoClass()` is a bag of methods, written in [concise method
```
Comment (dckc) [inline main/guides/zoe/contract-details.md:70]:
```suggestion
The methods argument of `exoClass()` is a collection of methods, written in [concise method
```

or "... record of methods ..."

File: main/guides/zoe/contract-details.md [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+syntax](https://github.com/Agoric/agoric-sdk/wiki/Arrow-Function-Style#far-classes-do-not-use-arrow-function-style).
+`exoClass()` defines a single facet.
+
+The methodKit parameter to `exoClassKit` is a record of labelled bags of methods. Each label defines a facet
```
Comment (dckc) [inline main/guides/zoe/contract-details.md:74]:
```suggestion
The methodKit parameter to `exoClassKit` is a record where each value is a collection of methods. Each property defines a facet
```

File: main/guides/zoe/contract-details.md [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+need. This means putting relevant state in
+[Baggage](http://localhost:5173/guides/zoe/contract-upgrade.html#baggage), and ensuring that reachable objects
+that will be accessible to users have an identity that can be maintained as the behavior changes with contract
+upgrades.
+
+We use zone.exo(), zone.exoClass(), and zone.exoClassKit() to define durable objects.
+
+[Zone](http://localhost:5173/glossary/#zone) provides an interface for defining objects and classes that
+supports both ephemeral objects (allocated on the heap), and durable objects that can persist and that
+SwingSet will page in or out on demand.
```
Comment (dckc) [inline main/guides/zoe/contract-details.md:24]: Is "SwingSet" in the reader's vocabulary at this point? Link to glossary entry?

File: main/guides/zoe/contract-details.md [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+  methodB: M.callWhen(M.await(paramsBShape)).returns(resultBShape),
+}
+```
+
+`M.call()` verifies that all parameters match the guard before passing them through to the
+method. `M.callWhen(M.await(paramsBGuard))` awaits the resolution of the parameter, and then verifies that the
+result matches before invoking the method. When a guard is written this latter way, the method doesn't have to
+be `async`. In both cases, the method doesn't have to check for compliance with the guard.
+
+[Shapes can specify](https://endojs.github.io/endo/interfaces/_endo_patterns.PatternMatchers.html) simple
```
Comment (dckc) [inline main/guides/zoe/contract-details.md:103]: I made a patterns cheat-sheet at https://docs.agoric.com/reference/zoe-api/zoe-contract-facet.html#proposal-shapes, but that context is specialized to proposal shapes. hm.

Perhaps using glossary entries for Pattern and Shape would help. Perhaps I should do that in a follow-up PR.

**Review by dckc — CHANGES_REQUESTED** [review]
Fixing the links is important.

The rest of the comments are just suggestions.

File: main/guides/zoe/contract-details.md (hunk 5 above)
Comment (Chris-Hibbert) [reply main/guides/zoe/contract-details.md]: done

File: main/guides/zoe/contract-details.md [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+We use zone.exo(), zone.exoClass(), and zone.exoClassKit() to define durable objects.
+
+[Zone](http://localhost:5173/glossary/#zone) provides an interface for defining objects and classes that
+supports both ephemeral objects (allocated on the heap), and durable objects that can persist and that
+SwingSet will page in or out on demand.
+
+Our persistent objects are designed to encapsulate their state, and can present different facets to different
+clients to distinguish different aspects of authority. `zone.exoClass()` defines a kind with a single facet,
+while `zone.exoClassKit()` defines a kind with multiple facets. (Here are some relevant [naming
+conventions](https://docs.agoric.com/guides/ertp/#method-naming-structure).)
```
[the remark this answers is not in the document]
Comment (Chris-Hibbert) [reply main/guides/zoe/contract-details.md]: done

File: main/guides/zoe/contract-details.md (hunk 2 above)
Comment (Chris-Hibbert) [reply main/guides/zoe/contract-details.md]: done

File: main/guides/zoe/contract-details.md (hunk 4 above)
Comment (Chris-Hibbert) [reply main/guides/zoe/contract-details.md]: enhanced

File: main/guides/zoe/contract-details.md (hunk 3 above)
Comment (Chris-Hibbert) [reply main/guides/zoe/contract-details.md]: done. I like "record" here.

File: main/guides/zoe/contract-upgrade.md [hunk 8]
```diff
@@ -1,8 +1,12 @@
 # Contract Upgrade
 
-The result of starting a contract includes the right to upgrade the contract instance. A call to [E(zoe).startInstance(...)](/reference/zoe-api/zoe.md#e-zoe-startinstance-installation-issuerkeywordrecord-terms-privateargs) returns a record of several objects that represent different levels of access.
-The `publicFacet` and `creatorFacet` are defined by the contract.
-The `adminFacet` is defined by Zoe and includes methods to upgrade the contract.
+The return value when starting a contract includes a capability that conveys the right to upgrade
```
[the remark this answers is not in the document]
Comment (Chris-Hibbert) [reply main/guides/zoe/contract-upgrade.md]: okay

File: main/guides/zoe/contract-upgrade.md [hunk 9]
```diff
@@ -1,8 +1,12 @@
...
+the contract instance. A call to
+[E(zoe).startInstance(...)](/reference/zoe-api/zoe.md#e-zoe-startinstance-installation-issuerkeywordrecord-terms-privateargs)
+returns a record of several objects that carry different powers. The `publicFacet` and
```
[the remark this answers is not in the document]
Comment (Chris-Hibbert) [reply main/guides/zoe/contract-upgrade.md]: I don't like "levels of access".  "kit of facets" (with links) works for me.

File: main/guides/zoe/contract-upgrade.md [hunk 10]
```diff
@@ -60,16 +64,26 @@ There are a few requirements for the contract that differ from non-upgradable co
 
 ### Upgradable Declaration
 
-The new code bundle declares that it supports upgrade by exporting a `prepare` function in place of `start`.
+The new code bundle declares that it supports upgrade by including a `meta` record in addition to
+`start`. (We used to indicate upgradability by using `prepare` instead of `start`, but that
+approach is deprecated.)`
```
[the remark this answers is not in the document]
Comment (Chris-Hibbert) [reply main/guides/zoe/contract-upgrade.md]: done

File: main/guides/zoe/contract-upgrade.md [hunk 11]
```diff
@@ -83,13 +97,13 @@ if (!baggage.has('rooms')) {
 }
 ```
 
-The `provide` function supports a concise idiom for this find-or-create pattern:
+The `provide` function supports a concise idiom for this get-or-create pattern:
```
[the remark this answers is not in the document]
Comment (Chris-Hibbert) [reply main/guides/zoe/contract-upgrade.md]: okay.

File: main/guides/zoe/contract-upgrade.md [hunk 12]
```diff
@@ -60,16 +64,26 @@ There are a few requirements for the contract that differ from non-upgradable co
...
 
-The new code bundle declares that it supports upgrade by exporting a `prepare` function in place of `start`.
+The new code bundle declares that it supports upgrade by including a `meta` record in addition to
+`start`. (We used to indicate upgradability by using `prepare` instead of `start`, but that
+approach is deprecated.)`
 
-<<< @/../snippets/zoe/src/02b-state-durable.js#export-prepare
+`meta` is a record with any or all of `upgradability`, `customTermsShape`, and `privateArgsShape`
+defined. The latter two are optional
+[Patterns](https://endojs.github.io/endo/modules/_endo_patterns.html) restricting respectively
```
[the remark this answers is not in the document]
Comment (Chris-Hibbert) [reply main/guides/zoe/contract-upgrade.md:73]: I don't feel up to writing that glossary entry. I'll leave that to your hypothetical future PR.

**Review by dckc — APPROVED** [review]

**Pull request merged**